File: app/src/main/java/com/telolahy/towerofhanoi/scene/BaseScene.java
package com.telolahy.towerofhanoi.scene;

import com.telolahy.towerofhanoi.GameActivity;
import com.telolahy.towerofhanoi.manager.ResourcesManager;

import org.andengine.engine.camera.BoundCamera;
import org.andengine.entity.scene.Scene;
import org.andengine.opengl.vbo.VertexBufferObjectManager;

/**
 * Created by stephanohuguestelolahy on 11/4/14.
 */
public abstract class BaseScene extends Scene {

    protected ResourcesManager mResourcesManager;
    protected VertexBufferObjectManager mVertexBufferObjectManager;
    protected BoundCamera mCamera;
    protected GameActivity mActivity;

    //---------------------------------------------
    // CONSTRUCTOR
    //---------------------------------------------

    public BaseScene() {

        mResourcesManager = ResourcesManager.getInstance();
        mVertexBufferObjectManager = ResourcesManager.getInstance().vertexBufferObjectManager;
        mCamera = ResourcesManager.getInstance().camera;
        mActivity = ResourcesManager.getInstance().activity;

        createScene();
    }

    //---------------------------------------------
    // ABSTRACTION
    //---------------------------------------------

    public abstract void createScene();

    public abstract void disposeScene();

    public abstract void onBackKeyPressed();
}

File: app/src/main/java/com/telolahy/towerofhanoi/scene/BaseScene.kt
package com.telolahy.towerofhanoi.scene

import com.telolahy.towerofhanoi.MainActivity
import com.telolahy.towerofhanoi.manager.ResourcesManager
import org.andengine.engine.camera.BoundCamera
import org.andengine.entity.scene.Scene
import org.andengine.opengl.vbo.VertexBufferObjectManager

/**
 * Created by stephanohuguestelolahy on 11/4/14.
 */
abstract class BaseScene : Scene() {
  @JvmField
  protected var mResourcesManager: ResourcesManager = ResourcesManager.instance

  @JvmField
  protected var mVertexBufferObjectManager: VertexBufferObjectManager = ResourcesManager.instance.vertexBufferObjectManager

  @JvmField
  protected var mCamera: BoundCamera = ResourcesManager.instance.camera

  @JvmField
  protected var mActivity: MainActivity = ResourcesManager.instance.activity

  //---------------------------------------------
  // ABSTRACTION
  //---------------------------------------------
  abstract fun createScene()
  abstract fun disposeScene()
  abstract fun onBackKeyPressed()

  //---------------------------------------------
  // CONSTRUCTOR
  //---------------------------------------------
  init {
    createScene()
  }
}